cloudinit/config/cc_kernel_modules.py

```diff
@@ -0,0 +1,396 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+"""Kernel Modules"""
+import copy
+import re
+from array import array
+from logging import Logger
+from textwrap import dedent
+from typing import List
+
+from cloudinit import log as logging
+from cloudinit import subp, util
+from cloudinit.cloud import Cloud
+from cloudinit.config import Config
+from cloudinit.config.schema import MetaSchema, get_meta_doc
+from cloudinit.settings import PER_INSTANCE
+
+MODULE_DESCRIPTION = dedent(
+    """\
+Manages and enhances kernel modules on a systemd based system.
+This module is capable of loading kernel modules at boot as well as
+enhancing it with parameters.
+Beside applying settings during runtime it will also persist all
+settings in ``/etc/modules-load.d`` and ``/etc/modprobe.d``.
+"""
+)
+
+DISTROS = ["debian", "ubuntu"]
+
+meta: MetaSchema = {
+    "id": "cc_kernel_modules",
+    "name": "Kernel Modules",
+    "title": "Manage and enhance kernel modules",
+    "description": MODULE_DESCRIPTION,
+    "distros": DISTROS,
+    "frequency": PER_INSTANCE,
+    "activate_by_schema_keys": ["kernel_modules"],
+    "examples": [
+        dedent(
+            """\
+    kernel_modules:
+      - name: wireguard
+        load: true
+      - name: v4l2loopback
+        load: true
+        persist:
+          options: "devices=1 video_nr=20 card_label=fakecam exclusive_caps=1"
+      - name: zfs
+        persist:
+          blacklist: true
+    """
+        ),
+    ],
+}
+
+__doc__ = get_meta_doc(meta)
+
+LOG = logging.getLogger(__name__)
+NL = "\n"
+REQUIRED_KERNEL_MODULES_KEYS = frozenset(["name"])
+
+DEFAULT_CONFIG: dict = {
+    "km_cmd": {
+        "update": ["update-initramfs", "-u", "-k", "all"],
+        "unload": ["rmmod"],
+        "is_loaded": ["lsmod"],
+    },
```

Comment:
These commands are distribution-specific.  FreeBSD will probably have to use something like `kldload` NetBSD `modload`etc.

Maybe this is too early to think about distro-specific implementations...

What do think about hanging them this config off of `cloudinint.distro.Distro` as class attrs and methods:
 -  `kernel_module_cmds`  = {'load': ['"systemctl", "restart", "systemd-modules-load"], 'update' ... 'unload'.. 'is_loaded', 'persist'...}
 -   `kernel_module_fns` = {'load': ..., 'persist': ...}
 -  `def manage_kernel_modules`:  takes params `action: str, module_name: Optional[str]` if self.kernel_module_cmds.get(action) is None raise NotImplementedError. module_name could be optional for "load" action to  
 -  *BSD classes could set kernel_module_cmds = {} to ensure NotImplementedError is raised if called

Reply:
To be honest, I already had such an implementation in mind and tried to implement such functionality the early stages of development for this module. After some failing unit tests I overruled my plan as my main intention was to develop this module for Debian based systems. Nevertheless, I really like the idea of enhancing this module to become functional for other distros too. I am going to implement it. Stay tuned :)

Reply:
> These commands are distribution-specific. FreeBSD will probably have to use something like `kldload` NetBSD `modload`etc.
> 
> Maybe this is too early to think about distro-specific implementations...
> 
> What do think about hanging them this config off of `cloudinint.distro.Distro` as class attrs and methods:
> 
> * `kernel_module_cmds`  = {'load': ['"systemctl", "restart", "systemd-modules-load"], 'update' ... 'unload'.. 'is_loaded', 'persist'...}
> * `kernel_module_fns` = {'load': ..., 'persist': ...}
> * `def manage_kernel_modules`:  takes params `action: str, module_name: Optional[str]` if self.kernel_module_cmds.get(action) is None raise NotImplementedError. module_name could be optional for "load" action to
> * *BSD classes could set kernel_module_cmds = {} to ensure NotImplementedError is raised if called

Implemented improvements you requested. Concerning the distro-specific implementation I would be delighted if you could help me to implement it. I already tried creating the above methods and class attrs you mentioned but I can't get it to work. Maybe you could give me an more detailed and deep dive example? thx For example: what value does belong to "load" key from `kernel_module_fns` dict?

Reply:
@chifac08 sorry I missed the response here. I'm working on crafting the distro class specific extensions for both manage_kernel_modules. update_initramfs  and persist_module file paths. I've attached a patch suggestion this is work in progress so we can discuss further. I'll update this comment tomorrow with a more complete suggestion

The diff below is currently missing the following, but I will update this comment/diff with minimally config file specialization for distro classes.:
 * updated unittests to mock distro.manage_kernel_modules
 * added config file specialization for persisting module loads across rebootfor debian vs NotImplemented for non-debian-based systems
 * established new unit tests for the distro-class extensions 

```diff
diff --git a/cloudinit/config/cc_kernel_modules.py b/cloudinit/config/cc_kernel_modules.py
index 8eee08da1..07c5f285a 100644
--- a/cloudinit/config/cc_kernel_modules.py
+++ b/cloudinit/config/cc_kernel_modules.py
@@ -257,7 +257,7 @@ def reload_modules(cloud: Cloud):
 
     @raises RuntimeError
     """
-
+    LOG.debug("Reloading kernel modules")
     try:
         out = cloud.distro.manage_service("restart", "systemd-modules-load")
         if re.search("Failed", out.stdout.strip()):
@@ -268,32 +268,7 @@ def reload_modules(cloud: Cloud):
         ) from e
 
 
-def is_loaded(module_name: str) -> bool:
-    """
-    Checks if a kernel module is already loaded
-
-    @param module_name: name of kernel module
-
-    @return bool
-    """
-    loaded = False
-    try:
-        out = subp.subp(
-            DEFAULT_CONFIG["km_cmd"]["is_loaded"], capture=True, shell=True
-        )
-        if re.search("^" + module_name, out.stdout.strip()):
-            LOG.debug("Kernel module %s is loaded", module_name)
-            loaded = True
-    except subp.ProcessExecutionError as e:
-        util.logexc(
-            LOG,
-            f"Could not determine status of module {module_name}:{NL}{str(e)}",
-        )
-
-    return loaded
-
-
-def unload(modules: list):
+def unload(cloud: Cloud, modules: list):
     """Unloads a list of kernel module
 
     This function unloads a list of kernel modules.
@@ -305,28 +280,34 @@ def unload(modules: list):
     """
 
     cmd = copy.copy(DEFAULT_CONFIG["km_cmd"]["unload"])
+    loaded_modules = cloud.distro.manage_kernel_modules("list")
     for module in set(modules):
         cmd.append(module)
-        try:
-            if is_loaded(module):
-                LOG.debug("Unloading kernel module %s", module)
-                subp.subp(cmd)
-        except subp.ProcessExecutionError as e:
-            raise RuntimeError(
-                f"Could not unload kernel module {module}:{NL}{str(e)}"
-            ) from e
+        if re.search("^" + module_name + "$", loaded_modules):
+            try:
+                cloud.distro.manage_kernel_modules("unload", module)
+            except subp.ProcessExecutionError as e:
+                raise RuntimeError(
+                    f"Could not unload kernel module {module}:{NL}{str(e)}"
+                ) from e
 
 
-def update_initial_ramdisk():
+def update_initial_ramdisk(cloud: Cloud):
     """Update initramfs for all installed kernels
 
-    @param distro_cfg: Distro config
+    :param distro_cfg: Distro config
 
-    @raises RuntimeError
+    :raises: RuntimeError on command failure to update initramfs
+             NotImplementedError on lack of distro support
     """
-    LOG.debug("Update initramfs")
+    if not cloud.distro.update_initramfs_cmd:
+        raise NotImplementedError(
+            "Unable to update initramfs for %s. Kernel module changes will not"
+            " persist across reboot",
+            cloud.distro.name,
+        )
     try:
-        subp.subp(DEFAULT_CONFIG["km_cmd"]["update"])
+        subp.subp(cloud.distro.update_initramfs_cmd)
     except subp.ProcessExecutionError as e:
         raise RuntimeError(
             f"Failed to update initial ramdisk:{NL}{str(e)}"
@@ -378,12 +359,10 @@ def handle(
             enhance_module(module["name"], module["persist"], unload_modules)
 
     # rebuild initial ramdisk
-    update_initial_ramdisk()
+    update_initial_ramdisk(cloud)
 
     # Unload modules (blacklisted or 'load' is false)
-    log.info("Unloading kernel modules")
-    unload(unload_modules)
+    unload(cloud, unload_modules)
 
     # reload kernel modules
-    log.info("Reloading kernel modules")
     reload_modules(cloud)
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index 735a78327..23a9104f0 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -16,7 +16,7 @@ import stat
 import string
 import urllib.parse
 from io import StringIO
-from typing import Any, Mapping, MutableMapping, Optional, Type
+from typing import Any, List, Mapping, MutableMapping, Optional, Type
 
 from cloudinit import importer
 from cloudinit import log as logging
@@ -88,7 +88,14 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
     hostname_conf_fn = "/etc/hostname"
     tz_zone_dir = "/usr/share/zoneinfo"
     default_owner = "root:root"
-    init_cmd = ["service"]  # systemctl, service etc
+    init_cmd: List[str] = ["service"]  # systemctl, service etc
+
+    kernel_module_cmd_map: Mapping[str, str] = {
+        "list": ["lsmod"],
+        "load": ["insmod"],
+        "unload": ["rmmod"],
+    }
+    update_initramfs_cmd: List[str] = ["update-initramfs", "-u", "-k", "all"]
     renderer_configs: Mapping[str, MutableMapping[str, Any]] = {}
     _preferred_ntp_clients = None
     networking_cls: Type[Networking] = LinuxNetworking
@@ -907,6 +914,35 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
             args.append(message)
         return args
 
+    def manage_kernel_module(
+        self, action: str, kernel_module: Optional[str] = None
+    ):
+        """
+        Perform the requested load, unload, persist for a kernel module.
+
+        :param action: String of operation to perform, one of the following:
+            - load: load the kernel module in the running environment
+            - unload: remove kernel module from the running environment
+            - list: List all loaded kernel modules
+
+        :raises: NotImplementedError on distros without specific support
+                 ProcessExecutionError on failure on non-zero exit for command
+        """
+        if action not in self.kernel_module_cmd_map:
+            raise NotImplementedError(
+                "Unable to %s kernel module %s on %s. Not implemented.",
+                action,
+                kernel_module,
+                distro,
+            )
+        cmd = self.kernel_module_cmd_map[action]
+        if action != "list":
+            cmd += kernel_module
+            LOG.debug(
+                "%sing kernel module %s", action, kernel_module.capitalize()
+            )
+        return subp.subp(cmd, capture=True)
+
     def manage_service(self, action: str, service: str):
         """
         Perform the requested action on a service. This handles the common
diff --git a/cloudinit/distros/alpine.py b/cloudinit/distros/alpine.py
index 4a23fe079..4cb85b17f 100644
--- a/cloudinit/distros/alpine.py
+++ b/cloudinit/distros/alpine.py
@@ -24,6 +24,7 @@ class Distro(distros.Distro):
     pip_package_name = "py3-pip"
     locale_conf_fn = "/etc/profile.d/locale.sh"
     network_conf_fn = "/etc/network/interfaces"
+    update_initramfs_cmd = []  # TODO(define initramfs support)
     renderer_configs = {
         "eni": {"eni_path": network_conf_fn, "eni_header": NETWORK_FILE_HEADER}
     }
diff --git a/cloudinit/distros/arch.py b/cloudinit/distros/arch.py
index 2d5cfbf65..98ba1d74e 100644
--- a/cloudinit/distros/arch.py
+++ b/cloudinit/distros/arch.py
@@ -22,6 +22,7 @@ class Distro(distros.Distro):
     locale_gen_fn = "/etc/locale.gen"
     network_conf_dir = "/etc/netctl"
     init_cmd = ["systemctl"]  # init scripts
+    update_initramfs_cmd = []  # TODO(define mkinicpio support)
     renderer_configs = {
         "netplan": {
             "netplan_path": "/etc/netplan/50-cloud-init.yaml",
diff --git a/cloudinit/distros/bsd.py b/cloudinit/distros/bsd.py
index c4e1e15c6..d0188d18a 100644
--- a/cloudinit/distros/bsd.py
+++ b/cloudinit/distros/bsd.py
@@ -20,6 +20,9 @@ class BSD(distros.Distro):
     # poweroff.
     shutdown_options_map = {"halt": "-H", "poweroff": "-p", "reboot": "-r"}
 
+    kernel_module_cmd_map = {}  # TODO(define modunload/modload/modstat)
+    update_initramfs_cmd = []  # TODO(define initramfs support)
+
     # Set in BSD distro subclasses
     group_add_cmd_prefix: List[str] = []
     pkg_cmd_install_prefix: List[str] = []
diff --git a/cloudinit/distros/gentoo.py b/cloudinit/distros/gentoo.py
index 37217fe43..6b502fae8 100644
--- a/cloudinit/distros/gentoo.py
+++ b/cloudinit/distros/gentoo.py
@@ -22,6 +22,12 @@ class Distro(distros.Distro):
     network_conf_fn = "/etc/conf.d/net"
     hostname_conf_fn = "/etc/conf.d/hostname"
     init_cmd = ["rc-service"]  # init scripts
+    kernel_module_cmd_map = {
+        "is_loaded": ["lsmod"],
+        "load": ["modprobe"],
+        "unload": ["modprobe", "-r"],
+    }
+    update_initramfs = ["update-initramfs", "-u", "-k", "all"]
     default_locale = "en_US.UTF-8"
 
     # C.UTF8 makes sense to generate, but is not selected
diff --git a/cloudinit/distros/opensuse.py b/cloudinit/distros/opensuse.py
index 00ed15142..d4f21a460 100644
--- a/cloudinit/distros/opensuse.py
+++ b/cloudinit/distros/opensuse.py
@@ -25,6 +25,7 @@ class Distro(distros.Distro):
     systemd_hostname_conf_fn = "/etc/hostname"
     systemd_locale_conf_fn = "/etc/locale.conf"
     tz_local_fn = "/etc/localtime"
+    update_initramfs = []  # TODO(dracut support to regen initramfs)
     renderer_configs = {
         "sysconfig": {
             "control": "etc/sysconfig/network/config",
diff --git a/cloudinit/distros/photon.py b/cloudinit/distros/photon.py
index 14cefe90f..9dd6b7116 100644
--- a/cloudinit/distros/photon.py
+++ b/cloudinit/distros/photon.py
@@ -19,6 +19,10 @@ class Distro(distros.Distro):
     network_conf_dir = "/etc/systemd/network/"
     systemd_locale_conf_fn = "/etc/locale.conf"
     resolve_conf_fn = "/etc/systemd/resolved.conf"
+    kernel_module_cmd_map = {
+        "list": ["lsmod"],  # TODO(sort load and unload for kernel modules)
+    }
+    update_initramfs_cmd = []
 
     renderer_configs = {
         "networkd": {
diff --git a/cloudinit/distros/rhel.py b/cloudinit/distros/rhel.py
index 320f4ba1d..c52b48d52 100644
--- a/cloudinit/distros/rhel.py
+++ b/cloudinit/distros/rhel.py
@@ -36,6 +36,8 @@ class Distro(distros.Distro):
     network_script_tpl = "/etc/sysconfig/network-scripts/ifcfg-%s"
     tz_local_fn = "/etc/localtime"
     usr_lib_exec = "/usr/libexec"
+
+    update_initramfs_cmd = []  # TODO(dracut support to regen initramfs)
     renderer_configs = {
         "sysconfig": {
             "control": "etc/sysconfig/network",
diff --git a/tests/unittests/config/test_cc_kernel_modules.py b/tests/unittests/config/test_cc_kernel_modules.py
index 62a87e921..4bbf54064 100644
--- a/tests/unittests/config/test_cc_kernel_modules.py
+++ b/tests/unittests/config/test_cc_kernel_modules.py
@@ -124,8 +124,10 @@ class TestKernelModules(CiTestCase):
             "update-initramfs: execution error"
         )
 
+        distro = mock.MagicMock()  # No errors raised
+        mycloud = FakeCloud(distro)
         with self.assertRaises(RuntimeError) as context_mgr:
-            cc_kernel_modules.update_initial_ramdisk()
+            cc_kernel_modules.update_initial_ramdisk(mycloud)
         self.assertIn(
             "Failed to update initial ramdisk:\n"
             "Unexpected error while running command.\n"
```